lsm: remove compaction interface type

matklad opened the pull request:
Use (tree_id, level_b) index pair for "dynamic" dispatch instead

cb22 commented:
@matklad not sure if you saw the failure here - the fuzz needs to be changed too.

matklad committed:
lsm: CompactionInterface is replaced with (tree_id, level) pair

We need to "type-erase" compactions. Rather than introducing a separate
enum for this purpose, lets identify compaction by a tree id and a level,
leveraging existing `tree_for_id` infrastructure.

matklad commented on src/lsm/forest.zig:
```diff
@@ -921,6 +956,9 @@ fn CompactionPipelineType(comptime Forest: type, comptime Grid: type) type {
             const first_beat = compaction_beat == 0;
             const half_beat = compaction_beat == @divExact(constants.lsm_batch_multiple, 2);
 
+            if (self.forest == null) self.forest = forest;
```
Forest does `bar_apply_to_manifest` from `compact_callback`, so I figured it'd be simpler to just always have forest set. We still have `callback` to assert the proper lifecycle. 

Though, I do not like re-entrancy of `Forest.compact_callback`, would be good to fix that, but out of scope for this PR.

cb22 replied:
> Though, I do not like re-entrancy of Forest.compact_callback, would be good to fix that, but out of scope for this PR.

Yes - that was originally to get things working with the existing manifest compaction without having to touch that code too!

matklad commented on src/lsm/forest.zig:
```diff
@@ -1060,35 +1102,18 @@ fn CompactionPipelineType(comptime Forest: type, comptime Grid: type) type {
         // wouldn't compile for some reason.
         fn blip_callback(
-            compaction_interface_opaque: *anyopaque,
+            slot_opaque: *anyopaque,
```
Looks like just passing a pointer to a slot works here?

matklad commented on src/testing/cluster/manifest_checker.zig:
```diff
@@ -54,15 +54,16 @@ pub fn ManifestCheckerType(comptime Forest: type) type {
                 checksum_stream.add(std.mem.asBytes(&level));
 
-                inline for (Forest.tree_id_range.min..Forest.tree_id_range.max + 1) |tree_id| {
+                inline for (Forest.tree_id_range.min..Forest.tree_id_range.max + 1) |tree_id_u16| {
```
Intentionally avoiding `enum.values` here such that the verification code doesn't rely on the order of filds in an enum, and does a stupid numeric loop.

matklad commented:
ready for review @cb22 !

cb22 commented on src/lsm/forest.zig:
```diff
@@ -1268,11 +1332,26 @@ fn CompactionPipelineType(comptime Forest: type, comptime Grid: type) type {
 
                 // TODO: CompactionInterface internally stores a pointer to the real
```
Dead comment here

matklad replied:
Thanks, removed!

cb22 reviewed (commented):
LGTM - just a dangling old comment!
